src/mqtt_client.rs

```diff
@@ -365,27 +390,96 @@ where
                 Ok(())
             }
 
+            ReceivedPacket::PingResp => {
+                self.state.borrow_mut().pending_pingreq_time.take();
+                Ok(())
+            }
+
             _ => Err(Error::Unsupported),
         }
     }
 
+    fn open_socket(&mut self) -> Result<(), N::Error> {
+        let socket = self.network_stack.open(Mode::NonBlocking)?;
+        self.socket.borrow_mut().replace(socket);
+        Ok(())
+    }
+
     fn connect_socket(&mut self) -> Result<(), Error<N::Error>> {
+        // Don't try connecting more than once per second to avoid network spam if
+        // the broker port is closed.
+        if let Some(last_connect) = &self.last_socket_connect_time {
+            let now = self.clock.try_now().unwrap();
+            if now
+                .checked_duration_since(last_connect)
+                .and_then(|diff| Seconds::<u32>::try_from(diff).ok())
+                .map(|diff| diff < Seconds(1u32))
+                .unwrap_or(false)
+            {
+                return Err(Error::Disconnected);
+            }
+        }
+        self.last_socket_connect_time = Some(self.clock.try_now().unwrap());
+
+        // Connect to the broker's TCP port.
+        let addr = SocketAddr::new(self.state.borrow().broker, 1883);
+        info!("Connecting to {}", addr);
         let mut socket_ref = self.socket.borrow_mut();
         let socket = socket_ref.take().unwrap();
+        socket_ref.replace(self.network_stack.connect(socket, addr)?);
+        Ok(())
+    }
 
-        // Connect to the broker's TCP port with a new socket.
-        // TODO: Limit the time between connect attempts to prevent network spam.
-        let socket = self
-            .network_stack
-            .connect(socket, SocketAddr::new(self.state.borrow().broker, 1883))?;
-
-        // Store the new socket for future use.
-        socket_ref.replace(socket);
+    fn is_ping_request_due(&self) -> bool {
+        let s = self.state.borrow();
+        if s.keep_alive_interval == Seconds(0_u32) {
+            return false;
+        }
+        // Note(unwrap): We should have at least sent out the initial connect packet.
+        let last_write_time = &s.last_write_time.unwrap();
+        // On error (clock wrapped) be conservative and send out a keepalive ping.
+        let now = self.clock.try_now().unwrap();
+        now.checked_duration_since(last_write_time)
+            .and_then(|diff| Seconds::<u32>::try_from(diff).ok())
```

Comment:
Document why calling `ok()` is acceptable. Ditto on the `unwrap_or` below.

Reply:
Already documented three lines above.